src/concrete/ml/sklearn/base.py

```diff
@@ -1690,3 +1697,358 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
+                for idx, i in enumerate(indices):
+                    x[i] = v[idx]
+                return x
+
+            def mul_tlu(a, b):
+                """Matrix multiplication.
+
+                Args:
+                    a (numpy.ndarray): An encrypted array
+                    b (numpy.ndarray): An encrypted array
+
+                Returns:
+                    numpy.ndarray: The result of a * b
+                """
+                return a * b
+
+            comparisons = numpy.zeros(x.shape)
+            idx = numpy.arange(x.size) + fhe_zeros(x.shape)
+
+            n, k = x.size, self.n_neighbors
+            ln2n = int(numpy.ceil(numpy.log2(n)))
+
+            for t in range(ln2n - 1, -1, -1):
+                p = 2**t
+                r = 0
+                d = p
+
+                for bq in range(ln2n - 1, t - 1, -1):
+                    q = 2**bq
+                    range_i = numpy.array(
+                        [i for i in range(0, n - d) if i & p == r and comparisons[i] < k]
+                    )
+                    if len(range_i) == 0:
+                        continue
+
+                    a = gather1d(x, range_i)  # x[range_i]
+                    a_i = gather1d(idx, range_i)  # idx[range_i]
+                    b = gather1d(x, range_i + d)  # x[range_i + d]
+                    b_i = gather1d(idx, range_i + d)  # idx[range_i + d]
+
+                    diff = a - b
+                    sign = diff < 0
+
+                    max_x = a + numpy.maximum(0, b - a)
+                    x = scatter1d(x, a + b - max_x, range_i)  # x[range_i] = a + b - max_x
+                    x = scatter1d(x, max_x, range_i + d)  # x[range_i + d] = max_x
+
+                    max_idx = a_i + mul_tlu((b_i - a_i), sign)
+
+                    # idx[range_i] = a_i + b_i - max_idx
+                    idx = scatter1d(idx, a_i + b_i - max_idx, range_i)
+                    idx = scatter1d(idx, max_idx, range_i + d)  # idx[range_i + d] = max_idx
+
+                    comparisons[range_i + d] = comparisons[range_i + d] + 1
+
+                    d = q - p
+                    r = p
+
+            x = []
+            for i in range((self.n_neighbors)):
+                x.append(idx[i])
+            x = fhe_array(x)
+
+            assert x.shape[0] == self.n_neighbors
+
+            return x
+
+        # 1. Pairwise_euclidiean distance
+        distance_matrix = pairwise_euclidean_distance(q_X)
+
+        # sqr not done
+
+        # 2. Sorting args
+        sorted_args = topk_sorting(distance_matrix.flatten())
+
+        return sorted_args
+
+    def predict(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) -> numpy.ndarray:
+
+        X = check_array_and_assert(X)
+
+        y_preds = []
+        for query in X:
+            # Argsort
+            arg_sort = super().predict(query[None], fhe)
+            assert arg_sort.size == self.n_neighbors
+            # Majority vote
+            # pylint: disable=protected-access
```

Comment:
`disable-next` instead of `disable`, else I think it disables for the whole file !